### PR TITLE
Make GCS Client thread-safe.

### PR DESCRIPTION
**Pull request by micafan**
make RedisCallbackManager thread-safe.

## What do these changes do?

Make async execution of redis commands(RedisContext and RedisAsioClient) thread-safe for core worker.

@zhijunfu @kfstorm 

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Review by raulchen — COMMENTED** [review]
I seems that the wrapper class isn't necessary. Can we just put locks in `RedisContext::RunAsync`, `RedisContext::RunArgvAsync` and `RedisContext::SubscribeAsync`? Changes should be simpler that way.

**Comment by micafan** [comment]
> I seems that the wrapper class isn't necessary. Can we just put locks in `RedisContext::RunAsync`, `RedisContext::RunArgvAsync` and `RedisContext::SubscribeAsync`? Changes should be simpler that way.

Lock RedisContext is not enough. RedisAsioClient::handle_write should be locked too.

**Comment by stephanie-wang** [comment]
> > I seems that the wrapper class isn't necessary. Can we just put locks in `RedisContext::RunAsync`, `RedisContext::RunArgvAsync` and `RedisContext::SubscribeAsync`? Changes should be simpler that way.
> 
> Lock RedisContext is not enough. RedisAsioClient::handle_write should be locked too.

I also don't understand why we can't just lock `RedisContext`? As long as we lock any read/write/subscription calls, it should be okay, right?

Also, a couple things about this PR. This is a pretty big deal, since before now, we have not had any locks in the backend. Can you please explain why this is necessary? Also, before we merge this, we absolutely must have end-to-end performance regression tests since the GCS is on the critical path for many operations.

**Comment by micafan** [comment]
> > > I seems that the wrapper class isn't necessary. Can we just put locks in `RedisContext::RunAsync`, `RedisContext::RunArgvAsync` and `RedisContext::SubscribeAsync`? Changes should be simpler that way.
> > 
> > 
> > Lock RedisContext is not enough. RedisAsioClient::handle_write should be locked too.
> 
> I also don't understand why we can't just lock `RedisContext`? As long as we lock any read/write/subscription calls, it should be okay, right?
> 
> Also, a couple things about this PR. This is a pretty big deal, since before now, we have not had any locks in the backend. Can you please explain why this is necessary? Also, before we merge this, we absolutely must have end-to-end performance regression tests since the GCS is on the critical path for many operations.

1. The reason is:

- RedisAsioClient::handle_write will read from `redisContext's obuf` and write to socket, then modify  `redisContext's obuf` (running on a special thread). Please refer `hiredis` method `redisAsyncHandleWrite` and `redisBufferWrite` for more details.

- RedisContext::RunAsync will write to `redisContext's obuf` (might running on many threads in core worker). Please refer `hiredis` method `redisAsyncCommand` and `__redisAppendCommand` for more details.

2. Performance:

- `backend` has only one thread, there will be no lock competition. 

- C++ 11 have done greatly optimized for mutex. The locking cost is the same as atomic.

- If you still has concern, we can disable this lock by macro definition in backend.

**Comment by micafan** [comment]
> > > I seems that the wrapper class isn't necessary. Can we just put locks in `RedisContext::RunAsync`, `RedisContext::RunArgvAsync` and `RedisContext::SubscribeAsync`? Changes should be simpler that way.
> > 
> > 
> > Lock RedisContext is not enough. RedisAsioClient::handle_write should be locked too.
> 
> I also don't understand why we can't just lock `RedisContext`? As long as we lock any read/write/subscription calls, it should be okay, right?
> 
> Also, a couple things about this PR. This is a pretty big deal, since before now, we have not had any locks in the backend. Can you please explain why this is necessary? Also, before we merge this, we absolutely must have end-to-end performance regression tests since the GCS is on the critical path for many operations.

And also, this lock only re

File: src/ray/gcs/redis_async_context.cc
Comment (kfstorm) [inline src/ray/gcs/redis_async_context.cc]: Maybe add some comments about why the locks are necessary?

**Review by kfstorm — DISMISSED** [review]

File: src/ray/gcs/actor_state_accessor_test.cc
Comment (micafan) [reply src/ray/gcs/actor_state_accessor_test.cc]: The test data will be written to redis. The larger the amount of data, the longer the concurrent read and write of the redis buffer, so that it can be verified to a certain extent whether it is normal in the case of multithreading.

File: src/ray/gcs/redis_async_context.cc
Comment (micafan) [reply src/ray/gcs/redis_async_context.cc]: Sounds good.

File: src/ray/gcs/redis_async_context.h
Comment (raulchen) [inline src/ray/gcs/redis_async_context.h]:
```suggestion
  /// Get the raw `redisAsyncContext` pointer.
```

**Review by raulchen — APPROVED** [review]
LGTM

**Comment by stephanie-wang** [comment]
Please do not merge this PR yet. Can you first **explain why this is necessary**? Even if there is no performance overhead right now, we should avoid locks wherever possible.

**Comment by micafan** [comment]
> Please do not merge this PR yet. Can you first **explain why this is necessary**? Even if there is no performance overhead right now, we should avoid locks wherever possible.

Mainly for core worker.
GCS Client is also used in `core worker`. And `core worker` is a `multi-threaded` process, so it might call GCS Client in many threads. I have confirmed this to @zhijunfu.

**Comment by stephanie-wang** [comment]
> > Please do not merge this PR yet. Can you first **explain why this is necessary**? Even if there is no performance overhead right now, we should avoid locks wherever possible.
> 
> I agreed that we should avoid locks wherever possible.
> 
> Mainly for core worker.
> `GCS Client` is also used in `core worker`. And `core worker` is a `multi-threaded` process, so it might call` GCS Client’ methods` in many threads. I have confirmed this to @zhijunfu.

Can we just lock the GCS client inside the core worker then? This solution seems overly complicated and will affect many other parts of the backend. Please keep in mind that we should try to avoid modifying core code like this unless we all agree that it is necessary.

**Comment by micafan** [comment]
> > > Please do not merge this PR yet. Can you first **explain why this is necessary**? Even if there is no performance overhead right now, we should avoid locks wherever possible.
> > 
> > 
> > I agreed that we should avoid locks wherever possible.
> > Mainly for core worker.
> > `GCS Client` is also used in `core worker`. And `core worker` is a `multi-threaded` process, so it might call` GCS Client’ methods` in many threads. I have confirmed this to @zhijunfu.
> 
> Can we just lock the GCS client inside the core worker then? This solution seems overly complicated and will affect many other parts of the backend. Please keep in mind that we should try to avoid modifying core code like this unless we all agree that it is necessary.

It is very natural for the client to guarantee thread safety. It is not necessary to avoid reasonable locks at the core code. It is necessary to weigh the impact and necessity of the lock.

Can you explain why this solution `will affect many other parts of the backend`?

**Comment by stephanie-wang** [comment]
The GCS client is used very heavily by the raylet, so this will definitely affect the backend.

Sorry, I don't understand what you mean about "the impact and necessity of the lock".

I would recommend closing this PR and opening an issue to discuss the problem so that we can decide collectively how to deal with it. I still don't understand why we can't put the lock in the core worker instead of in the GCS client.

**Comment by micafan** [comment]
> The GCS client is used very heavily by the raylet, so this will definitely affect the backend.
> 
> Sorry, I don't understand what you mean about "the impact and necessity of the lock".
> 
> I would recommend closing this PR and opening an issue to discuss the problem so that we can decide collectively how to deal with it. I still don't understand why we can't put the lock in the core worker instead of in the GCS client.

In fact, only locking outside the GCS Client is no guarantee of thread safety.

**Comment by raulchen** [comment]
@stephanie-wang Sorry, I was about to write this explanation before merging.

Q: Why add a wrapper class, instead of adding the lock in `RedisContext`.

A: This is because the purpose of the wrapper class is to make `redisAsyncContext`-related commands (`redisvAsyncCommand` for example) thread-safe, and provide C++ style API. While  `RedisContext` is actually a group of `redisAsyncContext` for different purposes. And `redisAsyncContext` is used not only in `RedisContext`, but also in `RedisAsioClient`. 

Q: Why not lock the call sites of the gcs client in core worker?

A: If we do that, we'll have to remember to add locks at every call site. This would be too verbose, error-prone and easy to forget. Also, the scope of the lock will be more than necessary and potentially hurt perf. Thus, I believe it's better to provide thread-safe guarantee within the client (and it's also a common practice).

Regarding the perf impact for raylet, using a lock in a single-threaded process should be negligible. And the scope of the lock is actually very small, see comment above `RedisAsyncContext`. 

Please let me know if you think this doesn't make sense. Thanks

**Comment by raulchen** [comment]
@stephanie-wang, I talked with @micafan offline. Locking call sites won't actually work. 
Because we have two threads that both need to be locked. One is the call site thread, the other is the thread of the `io_service` in `RedisAsioClient`.

File: src/ray/gcs/redis_async_context.h
Comment (stephanie-wang) [inline src/ray/gcs/redis_async_context.h]: Can you move all method definitions to the `.cc` file?

File: src/ray/gcs/redis_async_context.h
Comment (stephanie-wang) [inline src/ray/gcs/redis_async_context.h]: Methods should be documented with:

```
/// One-line description.
///
/// More details.
/// \param <name> <description>
/// \return <return description>
```

File: src/ray/gcs/redis_async_context.h
Comment (stephanie-wang) [inline src/ray/gcs/redis_async_context.h]: This comment about the lock should be moved down to where `mutex` is declared.

File: src/ray/gcs/redis_context.h
Comment (stephanie-wang) [inline src/ray/gcs/redis_context.h]: Document what data structures this mutex protects.

**Review by stephanie-wang — COMMENTED** [review]
Thanks, @micafan and @raulchen, I understand this better now. I left a few comments about documentation, but it looks good!

Can you also include a unit test for thread safety?

File: src/ray/gcs/redis_async_context.h
Comment (micafan) [reply src/ray/gcs/redis_async_context.h]: Sure. I am working on it.

File: src/ray/gcs/redis_context.h
Comment (micafan) [reply src/ray/gcs/redis_context.h]: Good point.

File: src/ray/gcs/redis_async_context.h
Comment (micafan) [reply src/ray/gcs/redis_async_context.h]: Done.

**Comment by micafan** [comment]
> Thanks, @micafan and @raulchen, I understand this better now. I left a few comments about documentation, but it looks good!
> 
> Can you also include a unit test for thread safety?

The native redis command is difficult to code, and the results are difficult to verify. Thread safety now is verified by unit testing of AccessStateAccessor.

**Comment by micafan** [comment]
GCSClient_perf
This is the performance test result.

**Comment by micafan** [comment]
This is the performance test result.
GCSClient_perf

**Comment by micafan** [comment]
GCSClient_perf

**Pull request merged**
